b2sdk/transfer/emerge/executor.py

```diff
@@ -0,0 +1,480 @@
+        return FileVersionInfoFactory.from_api_response(response)
+
+    def _get_unfinished_file_and_parts(
+        self, bucket_id, file_name, file_info, continue_large_file_id, emerge_parts_dict=None
+    ):
+        if 'listFiles' not in self.session.account_info.get_allowed()['capabilities']:
+            return None, {}
+
+        unfinished_file = None
+        finished_parts = {}
+
+        if continue_large_file_id is not None:
+            unfinished_file = self.services.large_file.get_unfinished_large_file(
+                bucket_id,
+                continue_large_file_id,
+                prefix=file_name,
+            )
+            if unfinished_file.file_info != file_info:
+                raise ValueError(
+                    'Cannot manually resume unfinished large file with different file_info'
+                )
+            finished_parts = {
+                part.part_number: part
+                for part in self.services.large_file.list_parts(continue_large_file_id)
+            }
+        elif 'plan_id' in file_info:
+            assert emerge_parts_dict is not None
+            unfinished_file, finished_parts = self._find_unfinished_file_by_plan_id(
+                bucket_id, file_name, file_info, emerge_parts_dict
+            )
+        elif emerge_parts_dict is not None:
+            unfinished_file, finished_parts = self._match_unfinished_file_if_possible(
+                bucket_id, file_name, file_info, emerge_parts_dict
+            )
+        return unfinished_file, finished_parts
+
+    def _find_unfinished_file_by_plan_id(self, bucket_id, file_name, file_info, emerge_parts_dict):
+        assert 'plan_id' in file_info
+        best_match_file = None
+        best_match_parts = {}
+        best_match_parts_len = 0
+        for file_ in self.services.large_file.list_unfinished_large_files(
+            bucket_id, prefix=file_name
+        ):
+            if file_.file_info == file_info:
+                finished_parts = {}
+                for part in self.services.large_file.list_parts(file_.file_id):
+                    emerge_part = emerge_parts_dict.get(part.part_number)
+                    if emerge_part is None:
+                        # large file with same `plan_id` has more parts than current plan
+                        # so we want to skip this large file because it is broken
+                        finished_parts = None
+                        break
+                    if emerge_part.is_hashable() and emerge_part.get_sha1() != part.content_sha1:
+                        continue  # auto-healing - `plan_id` matches but part.sha1 doesn't - so we reupload
+                    finished_parts[part.part_number] = part
+                if finished_parts is None:
+                    continue
+                finished_parts_len = len(finished_parts)
+                if best_match_file is None or finished_parts_len > best_match_parts_len:
```

Comment:
If we have input file `ABCDEF` and two existing sessions `ABCXYZ` and `ABC`, then we will choose to take over whatever is happening in `ABCXYZ` rather than using `ABC` which doesn't require discarding any parts. I think we should sort by the total contentLength (not amount) of parts that we'd need to upload and pick the one that is most the complete version, as if we have `B(5)C(5)` and `A(1000)B(5)`, we'll choose to reupload a gigabyte with the current solution while we could easily save 995MB... on the other hand... maybe this gigabyte is actually going to be a server-side copy and the 5MB we would need to really upload.

I think we need to discuss it on a call.

Reply:
slack convo:

> auto-healing large upload session streams used to be easy. Pick the matching file name with matching file info, find the first one with the highest number of completed parts and continue that
> 
> now parts are not of equal length
> 
> but some parts are server-side copied so if we have a choice of uploading 5MB of copying 1GB, copying is easier
> 
> and so now we'll need to check a plan of execution of finishing all of candidate sessions to pick the one on the top of a metric, where uploading data in bytes is to be strongly avoided and copying is also not good but not nearly as important as uploading
> 
> it's a complication in b2sdk implementation, but the result for the user will be simple: interrupted sessions finish quickly when restarted.

lets put better auto-healing in a separate PR